include/runtime_symbol_lookup.hpp

```diff
@@ -9,26 +9,85 @@
 #include "symbol_map.hpp"
 #include "symbol_table.hpp"
 
+#include "ddres_def.hpp"
+#include <array>
 #include <unordered_map>
 
 namespace ddprof {
 
 class RuntimeSymbolLookup {
 public:
   explicit RuntimeSymbolLookup(std::string_view path_to_proc)
-      : _path_to_proc(path_to_proc) {}
+      : _path_to_proc(path_to_proc), _cycle_counter(1) {}
+
+  SymbolIdx_t get_or_insert_jitdump(pid_t pid, ProcessAddress_t pc,
+                                    SymbolTable &symbol_table,
+                                    std::string_view jitdump_path);
+
   SymbolIdx_t get_or_insert(pid_t pid, ProcessAddress_t pc,
                             SymbolTable &symbol_table);
+
   void erase(pid_t pid) { _pid_map.erase(pid); }
 
+  void cycle() { ++_cycle_counter; }
+
 private:
-  using PidUnorderedMap = std::unordered_map<pid_t, SymbolMap>;
-  void fill_perfmap_from_file(int pid, SymbolMap &symbol_map,
-                              SymbolTable &symbol_table);
+  using FailedCycle = std::unordered_map<std::string, uint32_t>;
+  struct SymbolInfo {
+    SymbolMap _map;
+    FailedCycle _failed_cycle;
+  };
+  using PidUnorderedMap = std::unordered_map<pid_t, SymbolInfo>;
+
+  // Notes on JITDump strategy
+  //
+  // 1) Retrieve JITDump path
+  // Dso type will tell us that there is a JIT file.
+  // LLVM sources explain the logic about where we can find it. though we don't
+  // need that. The file is mmaped so we can get the path from there.
+  //
+  // We store in the DSOHdr the fact that we have a JITDump file for the pid.
+  //
+  // 2) Retrieve symbols
+  // Whenever we will come across the symbolisation of an unknown region,
+  // we use the runtime_symbol_lookup to check for existing symbols.
+  // If none are found, we parse the JITDump file if available.
+  // If not, we look for a perf-map file.
+  // Symbols are cached with the process's address.
+  //
+  DDRes fill_from_jitdump(std::string_view jitdump_path, pid_t pid,
+                          SymbolMap &symbol_map, SymbolTable &symbol_info);
+
+  DDRes fill_from_perfmap(int pid, SymbolMap &symbol_map,
+                          SymbolTable &symbol_table);
+
   FILE *perfmaps_open(int pid, const char *path_to_perfmap);
 
+  bool has_lookup_failure(const SymbolInfo &symbol_info,
+                          std::string_view path) const {
+    const auto it = symbol_info._failed_cycle.find(std::string(path));
+    if (it != symbol_info._failed_cycle.end()) {
+      // failure during this cycle
+      return it->second == _cycle_counter;
+    }
+    return false;
+  }
+
+  void flag_lookup_failure(SymbolInfo &symbol_info, std::string_view path) {
+    symbol_info._failed_cycle[std::string(path)] = _cycle_counter;
+  }
+
+  bool should_skip_symbol(const std::string &symbol);
```

Comment:
Agreed on the const. For the string_view, I'm not sure it has so much value for a private API.

Reply:
If the caller has a `std::string_view` and not a `std::string`, it will avoid the creation of a new `std::string`. In general, it's better to have `const std::string &` parameters replaced with `std::string_view` for this reason (a particular case when it's not possible is when we expect the input string to be null-terminated - it's not guaranteed for a `std::string_view`).

Reply:
I see some value in unifying the code (though not a lot of value in the current example). Thanks. fixed.